### PR TITLE
Businness/logic

### PR DESCRIPTION
**Pull request by KMosc**
- [x] validates_presence_of add
- [x] clearing code
- [ ] custom errors

### CODE REVIEW

**Comment by KMosc** [comment]
- [x] Validators add, I've deleted not needed if statements. I've added error handling where record does not exist.
image

**Comment by KMosc** [comment]
@wmaciejak @jandudulski Is it good? The tests?

File: Gemfile [hunk 1]
```diff
@@ -4,7 +4,6 @@ git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 ruby '3.0.0'
 gem 'redis'
 gem 'resque'
-gem 'rspec', '~> 3.0'
```
Comment (w0st) [inline Gemfile:7]: Why did you remove it from the project?

I suppose that you already have installed the RSpec on your machine but remember about potential teammates.

File: app/concepts/base/repository.rb [hunk 2]
```diff
@@ -20,7 +20,11 @@ def update_all(params)
           end
 
         def create(attrs)
-            adapter.create(attrs)
+            begin
+                adapter.create(attrs)
+            rescue ActiveRecord::RecordInvalid => error
+                ["Error": "NotFound"]
```
Comment (w0st) [inline app/concepts/base/repository.rb:26]: `NotFound`?

File: app/concepts/Payment/wrapper/use_case/buy.rb [hunk 3]
```diff
@@ -9,24 +9,29 @@ def initialize(wrapper)
...
+            payment(left,ticket_params, cinema_hall_id, seat, movie_id)
           end
         end
     
     private
 
     def validate?(id,password, left_Repository, right_Repository, seat, cinema_hall_id, movie_id)
-      @generateseats = UseCase::CinemaHalls::GenerateSeats.new(right_Repository).call(cinema_hall_id)
-      !password.blank? && left_Repository.where(id: id, cinema_hall_id: cinema_hall_id, movie_id: movie_id) && seat.in?(@generateseats) && (!left_Repository.exists?(:cinema_hall_id => cinema_hall_id, seat: seat, movie_id: movie_id)) && left_Repository.ticket_available?(cinema_hall_id, movie_id)
+      @available_seats = UseCase::CinemaHalls::GenerateSeats.new(right_Repository).call(cinema_hall_id)
```
Comment (w0st) [inline app/concepts/Payment/wrapper/use_case/buy.rb]: Do you really need to use here the instance variable (`@`)?

File: spec/requests/tickets_spec.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
+require 'rails_helper'
+
+RSpec.describe "Tickets"
+  describe "Fetch tickets" do
+    it 'works and return status 201' do
```
Comment (w0st) [inline spec/requests/tickets_spec.rb:5]: the name doesn't correspond with https://github.com/KMosc/APITRY/pull/13/files#diff-0d888e45cfd38a87fd4472cb377ce6785b8532af5b65011e6175ecb356025d74R7

**Review by w0st — COMMENTED** [review]
Let's call in and discuss few things. It should be easier for you

File: spec/requests/tickets_spec.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+              pad: "true", 
+              password: "XD", 
+              seat: "1B", 
+              ticket_desk_id: :ticket_desk_id, 
+              cinema_hall_id: :cinema_hall_id, 
+              movie_id: :movie_id
+              } 
+            } 
+        )
+      expect(response.status).to eq(422)
```
Comment (w0st) [inline spec/requests/tickets_spec.rb:27]: `422` (see: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422) means "no processable" (it's a kind of a bad request) - it's an error and definitely isn't what you expect.
```suggestion
      expect(response.status).to eq(422)
```

File: spec/requests/tickets_spec.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+      get("/ticket_desks/:id/cinema_halls/:cinema_hall_id/movies/:movie_id/ticket")
+      expect(response.status).to eq(200)
+    end
+end
+
+
+RSpec.describe "Tickets"
+  describe "create tickets" do
+    it 'works and return status 201' do
+      post("/ticket_desks/:id/cinema_halls/:cinema_hall_id/movies/:movie_id/ticket",
```
Comment (w0st) [inline spec/requests/tickets_spec.rb:15]: You haven't passed `id`, `cinema_hall_id`, and `movie_id` parameters for the URI.

File: spec/requests/ticket_desk_spec.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
+require 'rails_helper'
+
+RSpec.describe "Ticket Desks"
+  describe "Fetch ticket desks" do
+    it 'works and return status 201' do
+      get("/ticket_desks/:id/")
+      expect(response.status).to eq(204)
```
Comment (w0st) [inline spec/requests/ticket_desk_spec.rb:7]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/204

File: spec/requests/ticket_desk_spec.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
+require 'rails_helper'
+
+RSpec.describe "Ticket Desks"
+  describe "Fetch ticket desks" do
+    it 'works and return status 201' do
```
Comment (w0st) [inline spec/requests/ticket_desk_spec.rb:5]: It's not a 201 -> https://github.com/KMosc/APITRY/pull/13/files#diff-b9e388c6095c3865717ddc77d8bf7dda3f4b166e39e62349b24e65c033c7a66dR7

File: spec/requests/ticket_desk_spec.rb [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+      get("/ticket_desks/:id/")
+      expect(response.status).to eq(204)
+    end
+end
+
+
+RSpec.describe "Ticket Desks"
+  describe "create movies" do
+    it 'works and return status 201' do
```
Comment (w0st) [inline spec/requests/ticket_desk_spec.rb:14]:
```suggestion
    it 'works and return status 200' do
```?

File: spec/requests/cinema_halls_spec.rb [hunk 10]
```diff
@@ -0,0 +1,18 @@
+require 'rails_helper'
+
+RSpec.describe "Cinema Halls test"
+  describe "Fetch cinema halls" do
+    it 'works and return status 201' do
```
Comment (w0st) [inline spec/requests/cinema_halls_spec.rb:5]: The same here.

File: spec/requests/cinema_halls_spec.rb [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+      get("/ticket_desks/:id/cinema_halls")
+      expect(response.status).to eq(200)
+    end
+end
+
+
+RSpec.describe "Ticket Desks"
+  describe "create movies" do
+    it 'works and return status 201' do
```
Comment (w0st) [inline spec/requests/cinema_halls_spec.rb:14]: The same here.

File: spec/models/genre_spec.rb [hunk 12]
```diff
@@ -1,6 +1,4 @@
 require 'rails_helper'
 
 RSpec.describe Genre, type: :model do
```
Comment (w0st) [inline spec/models/genre_spec.rb:3]: I propose to delete the file if you don't have any unit tests for it.

File: spec/models/movie_spec.rb [hunk 13]
```diff
@@ -1,5 +1,4 @@
 require 'rails_helper'
 
 RSpec.describe Movie, type: :model do
-  pending "add some examples to (or delete) #{__FILE__}"
```
Comment (w0st) [inline spec/models/movie_spec.rb:4]: I propose to delete the file if you don't have any unit tests for it.

File: spec/models/cinema_hall_spec.rb [hunk 14]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe CinemaHall, type: :model do
```
Comment (w0st) [inline spec/models/cinema_hall_spec.rb:3]: I propose to delete the file if you don't have any unit tests for it.

File: config/environments/production.rb [hunk 15]
```diff
@@ -60,11 +60,12 @@
...
   # SMTP settings for gmail
   config.action_mailer.smtp_settings = {
-   :address              => "smtp.gmail.com",
-   :port                 => 587,
-   :user_name            => 'kamil.mazurek319@gmail.com',
-   :password             => ENV['railsmailpass'],
-   :enable_starttls_auto => true
+    port: 587,
+    address: 'smtp.gmail.com',
+    :user_name            => 'kamil.mazurek319@gmail.com',
```
Comment (w0st) [inline config/environments/production.rb:65]: You **shouldn't** keep credentials in the repository.
See:
https://blog.gitguardian.com/secrets-credentials-api-git/
https://pomeroy.me/2017/10/stop-committing-credentials-to-github/

File: app/models/ticket_desk.rb [hunk 16]
```diff
@@ -1,4 +1,4 @@
 class TicketDesk < ApplicationRecord
-    has_and_belongs_to_many :cinema_halls
+    has_and_belongs_to_many :cinema_halls, :inverse_of=>:ticket_desk
```
Comment (w0st) [inline app/models/ticket_desk.rb]: Keep consistency (see: https://github.com/KMosc/APITRY/pull/13/files#diff-103c7663225274279e2e159e0eec7a4fcaf32545109353486b5a95a18f3d3adaR12), please. It increases the readability of the code. Even better would be to start using a linter like [Rubocop](https://github.com/rubocop/rubocop)
```suggestion
    has_and_belongs_to_many :cinema_halls, :inverse_of=> :ticket_desk
```

File: app/models/ticket.rb [hunk 17]
```diff
@@ -2,12 +2,21 @@ class Ticket < ApplicationRecord
...
   validates :cinema_hall_id, numericality: { only_integer: true }
-  validates :movie_id, numericality: { only_integer: true }
+  validates :movie_id, :presence => true, :if => :exist?
 
   validates_uniqueness_of :seat, conditions: -> { where.not(movie_id: :movie_id, cinema_hall_id: :cinema_hall_id) }
+  
+  def exist?
+    Movie.exists?(:movie_id)
+  end
+
```
Comment (w0st) [inline app/models/ticket.rb:15]: Remove redunant empty lines, please. They generate noise in the file.

File: app/models/ticket.rb [hunk 18]
```diff
@@ -2,12 +2,21 @@ class Ticket < ApplicationRecord
   belongs_to :ticket_desk
   belongs_to :cinema_hall
   belongs_to :movie
+
+  validates_presence_of :cinema_hall
```
Comment (w0st) [inline app/models/ticket.rb:6]: Do you really need it? `belongs_to` (see: https://github.com/rails/rails/issues/34454) by default should validate `cinema_hall` :wink:

File: app/models/movie.rb [hunk 19]
```diff
@@ -1,5 +1,11 @@
 class Movie < ApplicationRecord
-  belongs_to :genre
-  has_and_belongs_to_many :clients
+  belongs_to :genre, foreign_key: 'genre_id'
```
Comment (w0st) [inline app/models/movie.rb:2]: Do you need to point the foreign key?
```suggestion
  belongs_to :genre
```

File: app/models/movie.rb [hunk 20]
```diff
@@ -1,5 +1,11 @@
...
+  has_and_belongs_to_many :tickets, :inverse_of=>:movie
   belongs_to :cinema_hall
+  validates_presence_of :cinema_hall
```
Comment (w0st) [inline app/models/movie.rb:5]: Do you really need it? belongs_to (see: rails/rails#34454) by default should validate cinema_hall wink

File: app/models/genre.rb [hunk 21]
```diff
@@ -1,3 +1,8 @@
 class Genre < ApplicationRecord
+    validates :id, :presence => true, :if => :exist?
+    validates :id, numericality: { only_integer: true }
 
+    def exist?
```
Comment (w0st) [inline app/models/genre.rb:5]: It shouldn't be an instance method! Read about the [class methods](https://medium.com/rubyinside/class-methods-in-ruby-a-thorough-review-and-why-i-define-them-using-class-self-af677ede9596) in Ruby, please.

File: app/models/cinema_hall.rb [hunk 22]
```diff
@@ -1,14 +1,15 @@
 class CinemaHall < ApplicationRecord
-has_and_belongs_to_many :tickets
-validate :id, :if => :exist?
+has_and_belongs_to_many :tickets, :inverse_of=>:cinema_hall
```
Comment (w0st) [inline app/models/cinema_hall.rb]: Why are you using the old Hash syntax? Do you use some old books/tutorials to learn? Ruby since 1.9 (14 years ago!) prefers the following syntax: 
`has_and_belongs_to_many :tickets, inverse_of: :cinema_hall`


Read more about it [here](https://launchschool.com/books/ruby/read/hashes), please.

File: app/controllers/ticket_desks_controller.rb [hunk 23]
```diff
@@ -8,7 +8,7 @@ def show
   end
 
   def create
-    repository=Repository::TicketDeskRepository.new(TicketDesk)
+    repository=Repository::TicketDeskRepository.new()
```
Comment (w0st) [inline app/controllers/ticket_desks_controller.rb]:
```suggestion
    repository=Repository::TicketDeskRepository.new
```

File: app/controllers/ticket_controller.rb [hunk 24]
```diff
@@ -25,18 +25,20 @@ def create
...
+        render json: ["log": "failed"], status: :unprocessable_entity
       end
-  end
+    end
       
   def bookin
     repo = Repository::TicketRepository.new
-    if !repo.make_reservation(ticket_params)
-      render json: ["log": "error"]
+    @reservation = repo.make_reservation(ticket_params)
```
Comment (w0st) [inline app/controllers/ticket_controller.rb]: Why are you using instance variable here? The local one is enough.

Read [this](https://stackoverflow.com/questions/10280801/normal-variables-vs-instance-variable-in-ruby-whats-the-difference), please.

File: app/concepts/ticket/use_case/taken.rb [hunk 25]
```diff
@@ -4,7 +4,8 @@ class Taken < UseCase::Base::Base
         def call(id, cinema_hall_id, movie_id)
           seats_taken=[]
           i = 0
-          repository.where(cinema_hall_id: cinema_hall_id, movie_id: movie_id).each do |ticket|
+          find = repository.where(cinema_hall_id: cinema_hall_id, movie_id: movie_id) rescue nil
```
Comment (w0st) [inline app/concepts/ticket/use_case/taken.rb:7]: Shouldn't be sth like that?
```suggestion
          tickets = repository.where(cinema_hall_id: cinema_hall_id, movie_id: movie_id) rescue nil
```

File: app/concepts/cinema_hall/use_case/generate_seats.rb [hunk 26]
```diff
@@ -5,11 +5,15 @@ module CinemaHalls
...
                 big_alphabet = ("A".."Z").to_a
-                @cinema_hall = repository.find_by(id)
-                volume =  @cinema_hall.read_attribute('volume')
-                seats=1.step(volume,1).to_a
-                for i in 0..volume-1
-                      seats[i] = "#{i/10+1}#{big_alphabet[i%10]}"
+                begin
+                    @cinema_hall = repository.find_by(id) 
+                    volume =  @cinema_hall.volume
+                    seats=1.step(volume,1).to_a
```
Comment (w0st) [inline app/concepts/cinema_hall/use_case/generate_seats.rb]:
```suggestion
                    seats=(1..volume).to_a
```

File: app/concepts/cinema_hall/use_case/generate_seats.rb [hunk 27]
```diff
@@ -5,11 +5,15 @@ module CinemaHalls
...
+                    for i in 0..volume-1
```
Comment (w0st) [inline app/concepts/cinema_hall/use_case/generate_seats.rb]:
```suggestion
                    seats.map.with_index do |seat, index|
```

File: app/concepts/ticket/use_case/taken.rb [hunk 28]
```diff
@@ -4,7 +4,8 @@ class Taken < UseCase::Base::Base
...
+          find.each do |ticket|
```
Comment (w0st) [inline app/concepts/ticket/use_case/taken.rb:8]: It can be simplified, let's discuss it on :calling:

**Review by w0st — CHANGES_REQUESTED** [review]

File: Gemfile (hunk 1 above)
Comment (GoodSingularity) [reply Gemfile:7]: @w0st There is rspec rails in development and test namespace in this Gemfile. I'm writing from other account

File: app/concepts/base/repository.rb (hunk 2 above)
Comment (GoodSingularity) [reply app/concepts/base/repository.rb:26]: Gonna replace it. I deleted most of my If statements.

File: app/concepts/Payment/wrapper/use_case/buy.rb (hunk 3 above)
Comment (GoodSingularity) [reply app/concepts/Payment/wrapper/use_case/buy.rb]: Well, I was thinking the same.
